crates/spawner/src/destroyer.rs

```diff
@@ -9,15 +9,51 @@ pub(crate) struct DestroyerPlugin;
 impl Plugin for DestroyerPlugin {
     fn build(&self, app: &mut App) {
         app.add_system(
+            find_dead
+                .in_base_set(GameSet::Update)
+                .run_if(in_state(AppState::InGame))
+                .in_set(SpawnerSet::Destroyer)
+                .in_set(DestroyerSet::FindDead),
+        )
+        .add_system(
             destroy
                 .in_base_set(GameSet::Update)
                 .run_if(in_state(AppState::InGame))
-                .in_set(SpawnerSet::Destroyer),
+                .after(DestroyerSet::SendData)
+                .in_set(SpawnerSet::Destroyer)
+                .in_set(DestroyerSet::DestroyDead),
         );
     }
 }
 
-fn destroy(
+#[derive(Debug, Clone, PartialEq, Eq, Hash, SystemSet)]
+enum DestroyerSet {
+    FindDead,
+    SendData,
+    DestroyDead,
+}
+
+/// Special marker trait which when used on a component the despawner sends an event with the data
+pub trait DespawnEventProducer {
+    type Data;
+
+    fn data(&self) -> Self::Data;
+}
+
+// Contains the entity and the data of the despawned component
+pub struct DespawnedComponents<T: DespawnEventProducer + Send + Sync>
+where
+    <T as DespawnEventProducer>::Data: Send + Sync,
```

Comment:
I would impose these trait bounds in `DespawnEventProducer` (`type Data: Send + Sync;`) because it does not make sense to have a `DespawnEventProducer` without the data being Send + Sync.